README.md
node-qiwi-api
================
[![npm](https://img.shields.io/npm/v/node-qiwi-api)](https://www.npmjs.com/package/node-qiwi-api)
[![npm](https://img.shields.io/npm/dt/node-qiwi-api.svg)](https://www.npmjs.com/package/node-qiwi-api)
[![npm](https://img.shields.io/npm/dm/node-qiwi-api.svg)](https://www.npmjs.com/package/node-qiwi-api)
----------------
Official [documentation](https://developer.qiwi.com/qiwiwallet/qiwicom_ru.html) for Qiwi api.

Get started
----------------
Firstly, get access token at [Qiwi](https://qiwi.com/api).
```
npm install node-qiwi-api
```
You can use callback and promise based api.

Initialise new wallet instance with your access token:
```js
var callbackQiwi = require('node-qiwi-api').callbackApi;
var asyncQiwi = require('node-qiwi-api').asyncApi;

var callbackWallet = new callbackQiwi(token);
var asyncWallet = new asyncQiwi(token);
```
Now you can get information about your wallet and make money transfers.

**Next examples presented for callback api.**

Identification data
----------------
```js
wallet.getIdentificationData(wallet, (err, data) => {
  if(err) {
    /*hanle error*/
  }
  console.log(data);
})
```
**wallet** - wallet number without plus (+) and with prefix (79991234567)

Identify wallet
----------------
```js
wallet.identifyWallet(wallet, requestOptions, (err, data) => {
  if(err) {
    /*hanle error*/
  }
  console.log(data);
})
```
**wallet** - wallet number without plus (+) and with prefix (79991234567)

requestOptions includes: 
* **birthDate** - Date of birth (YYYY-MM-DD)
* **firstName** - First name
* **middleName** - Middle name
* **lastName** - Last name
* **passport** - Serial and number of passport 
* **snils** - SNILS number
* **inn** - INN number
* **oms** - OMS number

Information about account
----------------
```js
wallet.getAccountInfo((err, data) => {
  if(err) {
    /*hanle error*/
  }
  console.log(data);
})
```

Get accounts
----------------
```js
wallet.getAccounts(wallet, (err, data) => {
  if(err) {
    /*hanle error*/
  }
  console.log(data);
})
```
**wallet** - wallet number without plus (+) and with prefix (79991234567)

Get possible account aliases
----------------
Return possible for creation at your wallet account aliases
```js
wallet.getPossibleAccountAliases(wallet, (err, data) => {
  if(err) {
    /*hanle error*/
  }
  console.log(data);
})
```
**wallet** - wallet number without plus (+) and with prefix (79991234567)

Create account
----------------
Create account in new currency
```js
wallet.createAccount(wallet, accountAlias, (err, data) => {
  if(err) {
    /*hanle error*/
  }
  console.log(data);
})
```
**wallet** - wallet number without plus (+) and with prefix (79991234567)
**accountAlias** - Account alias, possible values: *qw_wallet_rub, qw_wallet_kzt, qw_wallet_usd, qw_wallet_eur*

If call was successfull as *data* you will have *{ success: true }*, else *{ success: false }*

Set default account
----------------
Set default account for all operations
```js
wallet.setDefaultAccount(wallet, accountAlias, (err, data) => {
  if(err) {
    /*hanle error*/
  }
  console.log(data);
})
```
**wallet** - wallet number without plus (+) and with prefix (79991234567)
**accountAlias** - Account alias, possible values: *qw_wallet_rub, qw_wallet_kzt, qw_wallet_usd, qw_wallet_eur*

If call was successfull as *data* you will have *{ success: true }*, else *{ success: false }*

Operation history
----------------
```js
wallet.getOperationHistory(wallet, requestOptions, (err, data) => {
  if(err) {
    /*hanle error*/
  }
  console.log(data);
})
```
**wallet** - wallet number without plus (+) and with prefix (79991234567)

requestOptions includes: 
* **rows** - Amount of payments in response. Integer from 1 to 50. Required.
* **operation** - Operation type. ALL - all operations, IN - incoming only, OUT - outgoing only, QIWI_CARD - just payments by QIWI cards (QVC, QVP). Default - ALL
* **sources** - Payment source. Array of values. Allowable values: QW_RUB - ruble account of wallet, QW_USD - usd account of wallet, QW_EUR - euro account of wallet, CARD - added and not added to wallet cards, MK - account of mobile operator. If not presented, you will receive info from all sources
* **startDate** - Start date (YYYY-MM-ddThh:mm:ssZ). By default equals yesterday date. Use only with endDate
* **endDate** - End date (YYYY-MM-ddThh:mm:ssZ). By default equals current date. Use only with startDate
* **nextTxnDate** - Transaction date (YYYY-MM-ddThh:mm:ssZ), (see nextTxnDate in response). Use only with nextTxnId
* **nextTxnId** - Previous transaction number (see nextTxnId in response). Use only with nextTxnDate
Maximum interval between startDate and endDate - 90 days.

As example - information about 25 outgoing payments can be get by next way:
```js
wallet.getOperationHistory({rows: 25, operation: "OUT"}, (err, operations) => {
  /* some code */
})
```
Operations statistics
----------------
If you want to see statistics for summs of payments by period of time use this method. Example:
```js
wallet.getOperationStatistics(wallet, requestOptions, (err, data) => {
  if(err) {
    /*hanle error*/
  }
  console.log(data);
})
```
**wallet** - wallet number without plus (+) and with prefix (79991234567)

requestOptions: **operation, sources, startDate, endDate** - Parameters are similar to **getOperationHistory**.

Get transaction info
----------------
Example:
```js
wallet.getTransactionInfo(transactionId, (err, data) => {
  if(err) {
    /*hanle error*/
  }
  console.log(data);
})
```

Get transaction receipt
----------------
Example:
```js
wallet.getReceipt(transactionId, requestOptions, (err, data) => {
  if(err) {
    /*hanle error*/
  }
  console.log(data);
})
```
requestOptions includes: 
* **type** - Transaction type from getOperationHistory
* **format** - File format, see wallet.receiptFormat

Transfer to Qiwi wallet
----------------
```js
wallet.toWallet({ amount: '0.01', comment: 'test', account: '+79261234567' }, (err, data) => {
  if(err) {
    /* handle err*/
    }
  console.log(data);
})
```
* **amount** - Ammount of money
* **comment** - Commentary for payment.
* **account** - Receiver phone number (with international prefix)

Transfer to mobile phone
----------------
Similar to "transfer to qiwi wallet", but  number without international prefix:
```js
wallet.toMobilePhone({ amount: '0.01', comment: 'test', account: '9261234567' }, (err, data) => {
  if(err) {
    /* handle err*/
    }
  console.log(data);
})
```

Transfer to card
----------------
Similar to other transfers, but account is card number:
```js
wallet.toCard({ amount: '0.01', comment: 'test', account: '5213********0000' }, (err, data) => {
  if(err) {
    /* handle err*/
    }
  console.log(data);
})
```

Transfer to bank account
----------------
```js
wallet.toBank({ amount: '0.01', account: '5213********0000', account_type: '1', exp_date: 'MMYY' }, recipient, (err, data) => {
  if(err) {
    /* handle err*/
    }
  console.log(data);
})
```
* **ammount** - Ammount of money
* **account** - Receiver card/account number
* **account_type** - Type of bank identificator.
  * for Alfa-bank (Альфа-Банк) - 1
  * for OTP Bank (АО ОТП БАНК) - 1
  * for Rosselhozbank (АО РОССЕЛЬХОЗБАНК) - 5
  * for Russian Standard (Русский Стандарт) - 1
  * for VTB (ВТБ ПАО) - 5
  * for Promsvyazbank (Промсвязьбанк) - 7
  * for Sberbank (ПАО Сбербанк) - 5
  * for Renessans Credit (Ренессанс Кредит) - 1
  * for Moscow Credit Bank (ПАО Московский кредитный банк) - 5
* **exp_date** - Card expiration date (MMYY), as examlpe: 0218 - february 2018. Only for card transfer.
* **recipient** -
  * 464 - Alfa-bank (Альфа-Банка)
  * 466 - Tinkoff Bank (Тинькофф Банк)
  * 804 - OTP Bank (АО ОТП БАНК)
  * 810 - Rosselhozbank (АО РОССЕЛЬХОЗБАНК)
  * 815 - Russian Standard (Русский Стандарт)
  * 816 - VTB (ВТБ ПАО)
  * 821 - Promsvyazbank (Промсвязьбанк)
  * 870 - Sberbank (ПАО Сбербанк)
  * 881 - Renessans Credit (Ренессанс Кредит)
  * 1135 - Moscow Credit Bank (ПАО Московский кредитный банк)

Get currency exchange rates
----------------
```js
wallet.getCrossRates((err, data) => {
  if(err) {
    /* handle err*/
    }
  console.log(data);
})
```

Transfer by requisites
----------------
```js
wallet.toCard(requestOptions, (err, data) => {
  if(err) {
    /* handle err*/
    }
  console.log(data);
})
```
requestOptions includes: 
* **amount** - Amount of money
* **account** - Receiver account number
* **bankName** - Receiver bank name 
* **bik** - Receiver bank bik 
* **city** - Receiver city 
* **organizationName** - Receiver organization name 
* **inn** - Receiver organization inn 
* **kpp** - Receiver organization kpp 
* **nds** - 'НДС не облагается' or 'В т.ч. НДС'
* **purpose** - Purpose of payment
* **urgent** - 0 or 1. Urgent pay need 10 minutes to complete. Available from 9.00 AM to 8.30 PM (09.00 - 20.30) by Moscow time (GMT+3). Service fee - 25 rubles
* **senderName** - Sender name 
* **senderMiddleName** - Sender middle name
* **senderLastName** - Sender last name

Convert currency at you wallet
----------------
```js
wallet.convertCurrency({amount: 10, currency: wallet.currencyCode.KZT, account:'+79991234567'}, (err, data) => {
  if(err) {
    /* handle err*/
    }
  console.log(data);
})
```
requestOptions includes: 
* **account** - Phone number with plus and international prefix, as example +79991234567
* **currency** - Currency code - 3 number by ISO-4217 (see wallet.currencyCode)
* **amount** - Amount of money for calculate commission
* **comment** - Commentary for payment.


Get invoices
----------------
```js
wallet.getInvoices(requestOptions, (err, data) => {
  if(err) {
    /* handle err*/
    }
  console.log(data);
})
```
requestOptions includes:
* **rows** - Amount of payments in response. Integer from 1 to 50. Required.
* **from** - Date from
* **to** - Date to
* **nextId** - If present, method will return invoices from this id
* **nextDate** - If present, method will return invoices created before this time

Pay invoice
----------------
```js
wallet.payInvoice(invoiceId, currency, (err, data) => {
  if(err) {
    /* handle err*/
    }
  console.log(data);
})
```
* **invoiceId** - Invoice id from getInvoices (bills[].id)
* **currency** - Currency from getInvoices (bills[].sum.currency)

Cancel invoice
----------------
```js
wallet.cancelInvoice(invoiceId, (err, data) => {
  if(err) {
    /* handle err*/
    }
  console.log(data);
})
```
* **invoiceId** - Invoice id from getInvoices (bills[].id)

Check commission rates
----------------
```js
wallet.checkOnlineCommission(recipient, requestOptions, (err, data) => {
  if(err) {
    /* handle err*/
    }
  console.log(data);
})
```
**recipient** - Allowable values stored in wallet.recipients

requestOptions includes: 
* **account** - Phone number with international prefix or card/account number, as example 79991234567
* **amount** - Amount of money for calculate commission

Check operation commission
----------------
```js
wallet.checkCommission(recipient, (err, data) => {
  if(err) {
    /* handle err*/
    }
  console.log(data);
})
```
data.content.terms.commission.ranges[i]:
* **recipient** - Allowable values stored in wallet.recipients

Response contains:
* **bound** - Payment amount, starting from which the condition applies
* **rate** - Commission (absolute multiplier)
* **fixed** - Fixed amount of commission


CHANGELOG.md
2.2.0
=====================
Add
---------------------
* getAccountLimits - Get account limits

2.1.3
=====================
Update
---------------------
* Up axios version to v0.21.1

2.1.1
=====================
Add
---------------------
* toOther - payment method, that can help with making transfers to Yandex.Money, Steam etc.

Update
---------------------
* Fixed spelling of "ammount" in asyncApi.js and callbackApi.js

2.1.0
=====================
Add
---------------------
* getInvoices - get invoices
* payInvoice - pay invoice
* cancelInvoice - cancel not payed invoice

2.0.0
=====================
Add
---------------------
* getAccounts - get wallet accounts
* createAccount - create account in new currency by alias
* setDefaultAccount - set default account
* getPossibleAccountAliases - get possible account aliases
* getIdentificationData - get wallet identification data
* getCrossRates - get exchange rates
* getTransactionInfo - get information about transaction
* getReceipt - get receipt of transaction
* checkOnlineCommission - get information about future operation commission
* identifyWallet - identify wallet 
* convertCurrency - convert currency at wallet
* toRequisites - pay by requisites

Update
---------------------
* README.md
* recipients
* getOperationHistory - update path to v2
* getOperationStatistics - update path to v2
* getOperationHistory and getOperationStatistics - now accepts wallet into arguments and not calling getAccountInfo anymore

Rename
---------------------
* getOperationStats -> getOperationStatistics
* requestNewWebHookSecret -> getNewWebHookSecret

Remove
---------------------
* getBalance - now you can get same data from getAccounts

Infrastructure
---------------------
* Add async api
* Replace request with axios


node-qiwi-api.js
var request = require("request");

/**
 * Qiwi api
 * @param {string} token 
 * @link https://github.com/InsightAppDev/node-qiwi-api
 */
function Qiwi(token) {
    this.recipients = {
        qiwi: 99,
        visa_rus: 1963,
        mastercard_rus: 21013,
        visa_sng: 1960,
        mastercard_sng: 21012,
        mir: 31652,
        tinkoff: 466,
        alfa: 464,
        promsvyaz: 821,
        russkiy_standard: 815
    }
    this.token = token;
    this.headers = {
        'Accept': 'application/json',
        'content-type': 'application/json',
        'Authorization': 'Bearer ' + this.token
    }
    this.apiUri = "https://edge.qiwi.com/";

    this.txnType = {
        "IN": 0,
        "OUT": 1,
        "ALL": 2
    }

    /**
     * Возвращает информацию о текущем аккаунте
     * @param {function(err,data)} callback 
     */
    this.getAccountInfo = function (callback) {
        var options = {
            url: this.apiUri + 'person-profile/v1/profile/current',
            headers: this.headers,
            json: true
        };

        request.get(options, (error, response, body) => {
            callback(error, body);
        });
    }

    /**
     * Возвращает информацию о балансе аккаунта
     * @param {function(err,data)} callback 
     */
    this.getBalance = function (callback) {
        var options = {
            url: this.apiUri + 'funding-sources/v1/accounts/current',
            headers: this.headers,
            json: true
        };

        request.get(options, (error, response, body) => {
            callback(error, body);
        });
    }

    /**
     * Возвращает историю операций аккаунта
     * @param {{rows:number, operation:string, sources:string, startDate:Date, endDate:Date, nextTxnDate:Date, nextTxnId:number} requestOptions 
     * @param {function(err,data)} callback 
     */
    this.getOperationHistory = function (requestOptions, callback) {
        this.getAccountInfo((err, data) => {
            if (err != null) {
                return callback(err, null);
            }
            if (data == null) {
                return callback(new Error("Can not retrieve account info"), null);
            }

            var options = {
                url: this.apiUri + 'payment-history/v1/persons/' + data.authInfo.personId + '/payments',
                headers: this.headers,
                qs: {
                    rows: requestOptions.rows,
                    operation: requestOptions.operation,
                    sources: requestOptions.sources,
                    startDate: requestOptions.startDate,
                    endDate: requestOptions.endDate,
                    nextTxnDate: requestOptions.nextTxnDate,
                    nextTxnId: requestOptions.nextTxnId
                },
                json: true
            };

            request.get(options, (error, response) => {
                callback(error, response.body);
            });
        });
    }

    /**
     * Возвращает статистику по операциям
     * @param {{operation:string, sources:string, startDate:Date, endDate:Date}} requestOptions 
     * @param {function(err,data)} callback 
     */
    this.getOperationStats = function (requestOptions, callback) {
        this.getAccountInfo((err, data) => {
            if (err != null) {
                callback(err, null);
                return;
            }

            if (data == null) {
                callback(new Error("Can not retrieve account info"), null);
                return;
            }

            var options = {
                url: this.apiUri + 'payment-history/v1/persons/' + data.authInfo.personId + '/payments/total',
                headers: this.headers,
                qs: {
                    operation: requestOptions.operation,
                    sources: requestOptions.sources,
                    startDate: requestOptions.startDate,
                    endDate: requestOptions.endDate
                },
                json: true
            };

            request.get(options, (error, response, body) => {
                callback(error, body);
            });
        });
    }

    /**
     * Перевод средств на qiwi кошелек
     * @param {{amount:number, comment:string, account:string}} requestOptions 
     * @param {function(err,data)} callback 
     */
    this.toWallet = function (requestOptions, callback) {
        var options = {
            url: this.apiUri + 'sinap/terms/99/payments',
            headers: this.headers,
            body: {
                id: (1000 * Date.now()).toString(),
                sum: {
                    amount: requestOptions.amount,
                    currency: "643"
                },
                source: "account_643",
                paymentMethod: {
                    type: "Account",
                    accountId: "643"
                },
                comment: requestOptions.comment,
                fields: {
                    account: requestOptions.account
                }
            },
            json: true
        };

        request.post(options, (error, response, body) => {
            callback(error, body);
        });
    }

    /**
     * Перевод средств на мобильный телефон
     * @param {{amount:number, comment:string, account:string}} requestOptions 
     * @param {function(err,data)} callback 
     */
    this.toMobilePhone = function (requestOptions, callback) {
        detectOperator('7' + requestOptions.account, (err, data) => {
            if (err || data.code.value == "2") {
                throw new Error('Can\'t detect operator.');
            }
            else {
                var options = {
                    url: this.apiUri + 'sinap/terms/' + data.message + '/payments',
                    headers: this.headers,
                    body: {
                        id: (1000 * Date.now()).toString(),
                        sum: {
                            amount: requestOptions.amount,
                            currency: "643"
                        },
                        source: "account_643",
                        paymentMethod: {
                            type: "Account",
                            accountId: "643"
                        },
                        comment: requestOptions.comment,
                        fields: {
                            account: requestOptions.account
                        }
                    },
                    json: true
                };

                request.post(options, (error, response, body) => {
                    if (typeof body.code != 'undefined') error = body;
                    callback(error, body);
                });
            }
        });
    }

    /**
     * Перевод стредств на карту
     * @param {{amount:number, comment:string, account:string}} requestOptions 
     * @param {function(err,data)} callback 
     */
    this.toCard = function (requestOptions, callback) {
        this.detectCard(requestOptions.account, (err, data) => {
            if (err || data.code.value == "2") {
                callback(new Error('Wrong card number!', null));
            }
            else {
                var options = {
                    url: this.apiUri + 'sinap/terms/' + data.message + '/payments',
                    headers: this.headers,
                    body: {
                        id: (1000 * Date.now()).toString(),
                        sum: {
                            amount: requestOptions.amount,
                            currency: "643"
                        },
                        source: "account_643",
                        paymentMethod: {
                            type: "Account",
                            accountId: "643"
                        },
                        comment: requestOptions.comment,
                        fields: {
                            account: requestOptions.account
                        }
                    },
                    json: true
                };

                request.post(options, (error, response, body) => {
                    callback(error, body);
                });
            }
        });
    }

    /**
     * Перевод на банковский счет
     * @param {{ammount:number,account:string,account_type:number,exp_date:number}} requestOptions 
     * @param {number} recipient 
     * @param {function(err,data)} callback 
     */
    this.toBank = function (requestOptions, recipient, callback) {
        var options = {
            url: this.apiUri + 'sinap/terms/' + recipient + '/payments',
            headers: this.headers,
            body: {
                id: (1000 * Date.now()).toString(),
                sum: {
                    amount: requestOptions.amount,
                    currency: "643"
                },
                source: "account_643",
                paymentMethod: {
                    type: "Account",
                    accountId: "643"
                },
                comment: requestOptions.comment,
                fields: {
                    account: requestOptions.account,
                    account_type: requestOptions.account_type,
                    exp_date: requestOptions.exp_date
                }
            },
            json: true
        };

        request.post(options, (error, response, body) => {
            callback(error, body);
        });
    }

    /**
     * Проверяет значение комиссии по операции
     * @param {number} recipient идентификатор получателя (this.recipients)
     * @param {*} callback 
     */
    this.checkCommission = function (recipient, callback) {
        var options = {
            url: this.apiUri + 'sinap/providers/' + recipient + '/form',
            json: true
        };

        request.get(options, (error, response, body) => {
            callback(error, body);
        });
    }

    function detectOperator(phone, callback) {
        var options = {
            url: 'https://qiwi.com/mobile/detect.action',
            headers: this.headers,
            qs: {
                phone: phone
            },
            json: true
        };

        request.post(options, (error, response, body) => {
            callback(error, body);
        });
    }

    this.detectCard = function (cardNumber, callback) {
        var options = {
            url: 'https://qiwi.com/card/detect.action',
            headers: this.headers,
            qs: {
                cardNumber: cardNumber
            },
            json: true
        };

        request.post(options, (error, response, body) => {
            callback(error, body);
        });
    }

    /**
     * Добавляет вебхук по заданному url 
     * @param {string} url Url адрес вебхука
     * @param {number} txnType тип сообщений. 0 - "Входящие", 1 - "Исходящие". 2 - "Все". (this.txnType)
     */
    this.addWebHook = function (url, txnType, callback) {
        var options = {
            url: this.apiUri + 'payment-notifier/v1/hooks',
            headers: this.headers,
            qs: {
                hookType: 1,
                param: url,
                txnType: txnType
            },
            json: true
        };

        request.put(options, (error, response, body) => {
            callback(error, body);
        });
    }

    /**
     * Удаляет вебхук с заданным UUID
     * @param {string} hookId UUID вебхука
     * @param {*} callback 
     */
    this.removeWebHook = function (hookId, callback) {
        var options = {
            url: this.apiUri + 'payment-notifier/v1/hooks/' + hookId,
            headers: this.headers,
            json: true
        };

        request.delete(options, (error, response, body) => {
            callback(error, body);
        });
    }

    /**
     * Получает секретный ключ вебхука по UUID
     * @param {string} hookId UUID
     * @param {*} callback 
     */
    this.getWebHookSecret = function (hookId, callback) {
        var options = {
            url: this.apiUri + 'payment-notifier/v1/hooks/' + hookId + '/key',
            headers: this.headers,
            json: true
        };

        request.get(options, (error, response, body) => {
            callback(error, body);
        });
    }

    /**
     * Обновляет секретный ключ вебхука по UUID
     * @param {string} hookId UUID
     * @param {*} callback 
     */
    this.requestNewWebHookSecret = function (hookId, callback) {
        var options = {
            url: this.apiUri + 'payment-notifier/v1/hooks/' + hookId + '/newkey',
            headers: this.headers,
            json: true
        };

        request.post(options, (error, response, body) => {
            callback(error, body);
        });
    }

    /**
     * Возвращает данные об активнон вебхуке для данного кощелька(токена)
     * @param {*} callback
     */
    this.getActiveWebHook = function (callback) {
        var options = {
            url: this.apiUri + 'payment-notifier/v1/hooks/active',
            headers: this.headers,
            json: true
        };

        request.get(options, (error, response, body) => {
            callback(error, body);
        });
    }

    /**
     * Отправляет тестовый запрос на активный вебхук
     * @param {*} callback 
     */
    this.testActiveWebHook = function (callback) {
        var options = {
            url: this.apiUri + 'payment-notifier/v1/hooks/test',
            headers: this.headers,
            json: true
        };

        request.get(options, (error, response, body) => {
            callback(error, body);
        });
    }
}

module.exports = {
    Qiwi: Qiwi
}
